### PR TITLE
Increase the default pruning parameters

### PR DESCRIPTION
**Pull request by dvdplm**
Snapshots are taken at even 5000-blocks points in the chain (this increases the likelihood of hosts producing snapshots at the same point in the chain which increases the likelihood of a new node finding three identical snapshot manifests which is what is called a "trusted" snapshot (this essentially never ever happens in practice)), but it doesn't actually **start** taking the snapshot until 100 blocks later. This let's the chain "settle" and makes it less likely to take the snapshot over a reorg.

Now, the default pruning history of 64 means that if the state is larger than the default 32Mb of "pruning memory" we'll prune blocks older than current - 64. Apparently the size of blocks and their state is now often bigger than 32Mb and so we end up pruning the chain more aggressively than we did when the code was written. With 64 blocks of state we can't take a snapshot, we need 100.

This PR increases the default pruning history to 128 and the pruning memory to 48. This will enable more nodes to make snapshots and might reduce pruning aggressiveness a little bit (we prune at startup and after each block, in `commit_block()`.

Closes https://github.com/OpenEthereum/open-ethereum/issues/11557

### CODE REVIEW

File: ethcore/src/client/config.rs [hunk 1]
```diff
@@ -113,8 +113,8 @@ impl Default for ClientConfig {
 			verifier_type: VerifierType::Canon,
 			state_cache_size: 1 * mb,
 			jump_table_size: 1 * mb,
-			history: 64,
-			history_mem: 32 * mb,
+			history: 128,
+			history_mem: 48 * mb,
```
Comment (ordian) [inline ethcore/src/client/config.rs]: how is it calculated exactly? why not e.g. 64mb?

Comment (dvdplm) [reply ethcore/src/client/config.rs]: No calculation really. We know 32Mb is not enough (but it was until quite recently) so the code will attempt to prune quite often, and this is why snapshots now fail. Up until recently we probably stayed inside the 32Mb most of the time and so even with a 64 block history we'd actually have more than a 100 (enough to start a snapshot). By bumping to a little more we likely stay within the memory limit more often and don't have to attempt pruning.

**Comment by dvdplm** [comment]
@eduadiez Before merging this it would be nice to know it fixes snapshots on your end. Feel free to merge when you're confident it works. :)

**Comment by eduadiez** [comment]
After test this PR I get the same results: 
```
2020-03-12 08:11:07 UTC main INFO open_ethereum::run  Starting Open-Ethereum/v2.8.0-nightly-fb98ae60e-20200310/x86_64-unknown-linux-gnu/rustc1.41.1
2020-03-13 00:57:32 UTC Verifier #4 TRACE snapshot_watcher  Snapshot requested at block #9660000
2020-03-13 00:57:33 UTC Periodic Snapshot INFO snapshot::service  Taking snapshot at #9660000
2020-03-13 00:57:33 UTC Periodic Snapshot WARN snapshot  Tried to take a snapshot at #9660000 but the earliest available block is #9660037
2020-03-13 00:57:33 UTC Periodic Snapshot WARN ethcore_service::service  Failed to take snapshot at block #9660000: Snapshot error Attempted to create a snapshot at an old block while using a pruned database. Please re-run with the --pruning archive flag
```
It's weird that it only had 37 previous blocks like before even though this PR increase `history` and `history_mem`

I'm going to increase the memory to 64MB to see if there's a change.

Could we be missing some other parameter that limits it?

File: ethcore/src/client/config.rs (hunk 1 above)
Comment (ordian) [reply ethcore/src/client/config.rs]: What do you think of removing `history_mem` limit entirely? This parameter seems very fragile and unintuitive. Number of blocks should be enough.

Comment (dvdplm) [reply ethcore/src/client/config.rs]: I wonder if it's there to handle some kind of pathological block/series of blocks where the memory consumption went through the roof?

Comment (dvdplm) [reply ethcore/src/client/config.rs]: Running a node I noticed that the `journal_size` was ~55Mb, so maybe 64mb is the better default here.

**Comment by dvdplm** [comment]
I am currently taking a snapshot using this branch, so far so good.

**Comment by ordian** [comment]
@eduadiez did you try running the master or stable branch with `cli` params `--pruning-history` and `--pruning-memory` set to the values in this PR or just this branch?

File: ethcore/src/client/config.rs [hunk 2]
```diff
@@ -99,6 +99,9 @@ pub struct ClientConfig {
 impl Default for ClientConfig {
 	fn default() -> Self {
 		let mb = 1024 * 1024;
+		// Note/TODO: the defaults here are overridden by the command line
+		// argument parser and changes to defaults here must be performed "over
+		// there" as well (see `cli/mod.rs`).
```
Comment (ordian) [inline ethcore/src/client/config.rs]: hmm, I think the long-term solution would be to get rid of `cli/mod.rs` in favor of `structopt`

File: ethcore/snapshot/src/watcher.rs [hunk 3]
```diff
@@ -115,6 +115,9 @@ impl ChainNotify for Watcher {
 			// frequency of snapshots and ensures more snapshots are produced from similar points in
 			// the chain.
 			.filter(|num| num % self.period == 0 )
+			.inspect(|num| {
+				trace!(target: "snapshot_sync", "Candidate for taking a snapshot: #{}", num);
```
Comment (ordian) [inline ethcore/snapshot/src/watcher.rs]: how many candidates did you observe in practice?

**Comment by ordian** [comment]
@dvdplm what are your thoughts on https://github.com/openethereum/openethereum/pull/11558#discussion_r392132860? Do you think having `pruning-mem` is a good idea?

File: ethcore/src/client/config.rs (hunk 2 above)
Comment (dvdplm) [reply ethcore/src/client/config.rs]: YES. (Hence the 'todo' in my note.)

File: ethcore/snapshot/src/watcher.rs (hunk 3 above)
Comment (dvdplm) [reply ethcore/snapshot/src/watcher.rs]: 1

**Comment by dvdplm** [comment]
> @dvdplm what are your thoughts on [#11558 (comment)](https://github.com/openethereum/openethereum/pull/11558#discussion_r392132860)? Do you think having `--pruning-memory` is a good idea?

It comes from [here](https://github.com/openethereum/openethereum/issues/4090) and it seems like the idea was to provide a means to cap memory consumption – and not, as I hypothesized, as a countermeasure to some attack. I think it is plenty confusing to users to have two interacting settings to control pruning and I don't think exposing the `--pruning-memory` settings adds much value.

@eduadiez thoughts?

File: ethcore/src/client/config.rs (hunk 2 above)
Comment (dvdplm) [inline ethcore/src/client/config.rs]:
```suggestion
		// there" as well (see `cli/mod.rs` and https://github.com/openethereum/openethereum/issues/11574).
```

File: ethcore/src/client/config.rs (hunk 1 above)
Comment (tjayrush) [reply ethcore/src/client/config.rs]: > I wonder if it's there to handle some kind of pathological block/series of blocks where the memory consumption went through the roof?

There were some pretty pathological blocks during the Fall 2016 DDos attacks on Mainnet Ethereum. Not sure if that's relevant.

**Pull request merged**